Put data in the columns "Tumor" and "Allele Frequency" when downloading data

oplantalech opened the pull request:
Patients with multiple samples have the column `Tumor` which displays the sample number(s) where the alteration occurs. However, this column, and the column `Allele Frequency`, are empty in the exported tables. This PR fixes that by creating extra rows for those mutations. Also, the "Allele Counts" column have been rearranged. This changes do not alter the visualized tables in the Patient View.

For this table ([link to the Patient View](http://www.cbioportal.org/beta/case.do?cancer_study_id=lgg_ucsf_2014&case_id=P04#/patient?caseId=P04&studyId=lgg_ucsf_2014)):
![captura de pantalla 2017-07-21 a les 9 43 13](https://user-images.githubusercontent.com/20278013/28454165-7beceac4-6dfa-11e7-987f-92b851771ccb.png)

This is (part of) the download table:
![captura de pantalla 2017-07-21 a les 9 53 00](https://user-images.githubusercontent.com/20278013/28454173-841a768a-6dfa-11e7-9df4-19a47f9241ca.png)

alisman commented:
@oplantalech this is a fix for https://github.com/cBioPortal/cbioportal/issues/2672, correct?

oplantalech commented:
@alisman Yes! Thank you for referencing it!

inodb commented:
@oplantalech Thanks for submitting! The solution you submitted is different from what @pieterlukasse proposed. I think it would be nicer to have one row per sample if possible, instead of having extra columns for each sample. That would also get around the issue of having this weird `;` separated format for the variant/ref reads.

pieterlukasse commented:
@inodb before @oplantalech I reviewed the code with him and based on this we decided to go for the solution that is now being presented in this PR. I think either format works and has its advantages / disadvantages. For the "Allele Freq" column, which contains 1 value per sample, one possible future solution would be to also put this in multiple columns (e.g. " Allele Freq (sample 1), Alle Freq (sample 2), etc" ). 

We will also be presenting this to some end users to get their feedback. So the solution could still change based on that. Maybe you can also check with some users on your side?

inodb commented:
@pieterlukasse Gotcha. I think either users know how to do a groupby using R/python or they don't. For the latter one the proposed format in this PR is prolly better. But for any type of programmatic use I'd prefer the former.

inodb commented:
@pieterlukasse @oplantalech Another thing is that getting all the samples in the column dropdown seems a little much. Especially if you have over 10 samples. It would be nice to be able to see overlap between samples and sort. I think however we can just have this on the `Tumor` column. I think clicking on that should sort by (1) the number of samples with that mutation and (2) the id of the sample. If we do decide to add the extra column for the samples it's probably better if it is only in the csv

pieterlukasse commented:
I just saw this PR has not been updated to the latest version. In the latest version the solution is different than initially described. This can be found here: https://github.com/thehyve/cbioportal-frontend/tree/export_patient_tables.  @oplantalech can you archive this old code, and put the new code here? It is a bit confusing to discuss this PR and a different branch at the same time. We should all be looking at the same code.

onursumer commented on src/shared/components/mutationTable/column/SampleDownloadColumnFormatter.spec.tsx:
```diff
@@ -0,0 +1,79 @@
+import sinon from 'sinon';
+
+/**
+ * @author Selcuk Onur Sumer
```
I don't think I am the author of this completely new file :)

pieterlukasse replied:
@oplantalech if there is a new file, please change `@author`

onursumer commented on src/shared/components/mutationTable/column/SampleDownloadColumnFormatter.tsx:
```diff
@@ -0,0 +1,111 @@
+import styles from "./sample.module.scss";
+
+/**
+ * @author Selcuk Onur Sumer
```
Do we really need to duplicate this file too?

oplantalech commented:
@onursumer @inodb @alisman @pieterlukasse Sorry for the duplication of the branch. Now the PR is up to date.

inodb commented on src/pages/patientView/mutation/column/TumorColumnFormatter.tsx:
```diff
@@ -67,4 +66,12 @@ export default class TumorColumnFormatter {
+        if (data) {
+            result = data.sampleId;
+        }
+        return result;
```
This is more readable imo as `return data? data.sampleId : '';`

alisman commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -17,18 +17,23 @@ import { If } from 'react-if';
 import {SortMetric} from "../../lib/ISortMetric";
 import {IMobXApplicationDataStore, SimpleMobXApplicationDataStore} from "../../lib/IMobXApplicationDataStore";
 
+import {shallow, mount, ReactWrapper} from 'enzyme';
```
why is this imported here?

alisman commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -17,18 +17,23 @@ import { If } from 'react-if';
     sortBy?:((data:T)=>(number|null)) | ((data:T)=>(string|null)) | ((data:T)=>(number|null)[]) | ((data:T)=>(string|null)[]);
     render:(data:T)=>JSX.Element;
-    download?:(data:T)=>string;
+    download?:(data:T, sample?:TChild)=>string;
```
it seems we are still referencing sample here.   LazyMobxTable should not be aware of sample.

onursumer commented on src/pages/patientView/mutation/column/AlleleFreqColumnFormatter.spec.tsx:
```diff
@@ -52,4 +54,11 @@ describe('AlleleFreqColumnFormatter', () => {
+            tumorAltCount:1,
+            tumorRefCount:10
+        });
+        assert.deepEqual(AlleleFreqColumnFormatter.getFrequency(mutation), "0.09090909090909091");
```
Why do we need `deepEqual` for a string comparison?

onursumer commented on src/shared/components/lazyMobXTable/LazyMobXTable.spec.tsx:
```diff
@@ -149,40 +149,47 @@ describe('LazyMobXTable', ()=>{
         render:(d:any)=>(<span>{d.name}</span>),
-        download:(d:any)=>d.name,
+        download:(d:any, s:any)=>s.name,
+        toDownload: true,
```
I don't think we need `toDownload`, instead we can check if `download` is defined or not

onursumer commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -17,18 +17,23 @@ import { If } from 'react-if';
     tooltip?:JSX.Element;
     defaultSortDirection?:SortDirection;
     togglable?:boolean;
+    toDownload?:boolean;
```
I don't think we need toDownload, instead we can check if download is defined or not

pieterlukasse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -208,29 +213,33 @@ class LazyMobXTableStore<T> {
+                    this.columns.forEach((column:Column<T>) => {
+                        if (column.toDownload) {
+                            if (column.download) {
+                                rowDownloadData.push(column.download(rowData, sample));
```
don' t use "sample" , but a generic variable name

pieterlukasse commented on src/shared/components/lazyMobXTable/LazyMobXTable.spec.tsx:
```diff
@@ -975,17 +982,17 @@ describe('LazyMobXTable', ()=>{
-                "1,6,kdfjpo,,1HELLO123456,,\r\n"+
-                "4,90,zkzxc,,4HELLO123456,,\r\n" +
-                "2,null,null,,2HELLO123456,,\r\n");
-        });
```
from @inodb : please add specific test for multiple samples

pieterlukasse commented:
@oplantalech as @inodb mentioned: please also test other places where the table is used to make sure the new code does not break anything there.

pieterlukasse commented on src/shared/components/lazyMobXTable/LazyMobXTable.spec.tsx:
```diff
@@ -149,40 +149,47 @@ describe('LazyMobXTable', ()=>{
         render:(d:any)=>(<span>{d.num}</span>),
-        download:(d:any)=>d.num+'',
+        download:(d:any, s:any)=>s.num+'',
+        toDownload: true,
```
decision: please remove this "toDownload" feature

pieterlukasse commented on src/shared/components/lazyMobXTable/LazyMobXTable.spec.tsx:
```diff
@@ -959,13 +966,13 @@ describe('LazyMobXTable', ()=>{
         });
         it("gives one row of data when theres one row. data given for every column, including hidden, and without download def'n. if no data, gives empty string for that cell.", ()=>{
-            let table = mount(<Table columns={columns} data={[data[0]]}/>);
+            let table = mount(<Table columns={columns} data={[[data[0]]]}/>);
```
decision: revert all changes to test

pieterlukasse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -208,29 +213,33 @@ class LazyMobXTableStore<T> {
-            });
-
-            tableDownloadData.push(rowDownloadData);
+            }
```
decision: use extra "else"  here for when `rowData` is not an Array (which is indeed not always the case).

inodb commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -208,29 +213,33 @@ class LazyMobXTableStore<T> {
     @computed public get downloadData()
```
@onursumer @alisman What are your thought on changing this function this way? Should we override this function in the `MutationTable` and `CNATableComponent`? Or should we pass some function for downloading? I think eventually we want to be able to support multiple download formats from mutation table, e.g. MAF/VCF/TXT. But for now maybe better to support this proposed MAF-like format with one row per sample.

alisman replied:
Hi guys, I think we should pass a function for downloading.  Formatting the data for download seems very specific to the nature of the particular data being displayed and can't be generalized here.   @adamabeshouse would be a good person to discuss this with.   Is it important that the downloaded data reflect ordering and filtering state or can we just deliver it as a dump given that someone will most likely be using another tool to manipulate the data?

inodb replied:
I think use case depends on user. I would reckon that the folks who want an excel style output format, might want to get the filtered state, so they don't have to do any post processing. Bioinformaticians are probably happy with a raw data dump.

adamabeshouse replied:
yeah the table component should be completely agnostic about the form of the data, download should take the rowData and the table shouldn't have any opinion on whether its an array or not.

oplantalech replied:
@adamabeshouse I think it is not possible to move all the changes I made to the download functions: with this extra loop, we are creating new 'extra' rows (one row per every sample instead of one row per every mutation/alteration). Currently, the `push` function only adds one string to the row, according to the order of the columns. Therefore, if I want to maintain this 'extra' rows but remove the extra loop I introduced, I could make the `download` functions to return arrays instead of strings. However, these arrays should be parsed differently, and this should be done in LazyMobXTable, and I believe that's not what it is desired.

adamabeshouse replied:
I see now, sorry I didn't think about it carefully enough, you're right. I actually think having `download` return an array would be an elegant way to handle it.

pieterlukasse replied:
I think the current solution, which @adamabeshouse summarized well, could be the default behavior. Perhaps other (future) download formats could be implemented by overriding this default function? So by the time we need more specific download formats for the patient view, we could have a new type, e.g. PatientViewTable, which extends LazyMobXTable with its own extra download formats. Is such a scenario feasible with the current code structure?

adamabeshouse replied:
@pieterlukasse actually I don't think what I said is the same - the current solution is, if the `rowData` is an array, loop through it and call `download` on each of its elements, then combine them. What I'm suggesting is, call `download` on `rowData`, always, and deal with whatever it returns.

I think a good way to handle the multiple desired download formats could be: put a prop on `LazyMobXTable` called `download` which takes the result of calling each of the column `download` functions on all the data, and decides how to combine them.

So each row gives an array of either `string` or `string[]`, i.e. `(string|string[])[]`, thus the input to `LazyMobXTable.props.download` would be of type `(string|string[])[][]`

oplantalech commented:
Checking if the changes in this PR break the functionality of the other cBioPortal `LazyMobXTables`:

- Datasets: I could not check it, but columns have the download option set to 'false': 
```
columns={
    [
        {
            name:'Name',
            type: 'name',
            render:(data:IDataTableRow)=> <CancerStudyCell studyId={data.studyId} name={data.name}/>,
            filter:(data:any, filterString:string, filterStringUpper:string) => {
                return data.name.toUpperCase().indexOf(filterStringUpper) > -1;
             }

        },
        {name:'', sortBy:false, togglable:false, download: false, type:'download', render:(data:IDataTableRow)=> {
            const download = this.state.downloadable.indexOf(data.studyId) > -1;
            return (
                <a className="dataset-table-download-link" style={download ? {display:'block'} : {display:'none'}}
                   href={'https://media.githubusercontent.com/media/cBioPortal/datahub/master/public/' + data.studyId + '.tar.gz'} download>
                    <i className='fa fa-download'/>
                </a>
            );
        }},
```

- Clinical information tables: when you click the download button, they are empty in the PR and also in: http://www.cbioportal.org/beta/case.do#/patient?caseId=P04&studyId=lgg_ucsf_2014&tab=clinicalDataTab

- Cosmic tables in Patient View and Result's Mutation tab: no changes (as expected, since there is no download option for those tables).

- Result's Mutation tab: in beta, the column "Sample ID" in the download table is full, while in the PR the "Sample ID" column is empty. However, if I switch to rc, the "Sample ID" columns are still empty. So I believe the code used to fill the "Sample ID" column is outside this repository.

Sjoerd-van-Hagen commented:
@jjgao Hi JJ, do you think this is ready to be merged soon?

adamabeshouse commented on src/pages/patientView/copyNumberAlterations/CopyNumberTableWrapper.tsx:
```diff
@@ -60,6 +60,8 @@ export default class CopyNumberTableWrapper extends React.Component<ICopyNumberT
                 render:(d:DiscreteCopyNumberData[])=>TumorColumnFormatter.renderFunction(d, this.props.sampleManager),
                 sortBy:(d:DiscreteCopyNumberData[])=>TumorColumnFormatter.getSortValue(d, this.props.sampleManager),
+                download: (d:DiscreteCopyNumberData[], s:DiscreteCopyNumberData)=>TumorColumnFormatter.getSample(s),
+                toDownload: true,
```
what is the purpose of `toDownload`? can it not be inferred from the presence of a value for `download`?

adamabeshouse commented on src/pages/patientView/mutation/column/AlleleFreqColumnFormatter.tsx:
```diff
@@ -142,4 +142,15 @@ export default class AlleleFreqColumnFormatter {
         });
     }
+
+    public static getFrequency(data:Mutation): string {
```
more specific name would be good, `getTumorVariantAlleleFrequency` maybe?

adamabeshouse commented on src/pages/patientView/mutation/column/TumorColumnFormatter.tsx:
```diff
@@ -67,4 +66,12 @@ export default class TumorColumnFormatter {
         }, {});
     }
+
+    public static getSample(data:{sampleId:string}): string {
```
`getSampleId`, because there are `Sample` objects so this is a bit confusing

adamabeshouse commented on src/shared/components/mutationTable/column/AlleleCountColumnFormatter.tsx:
```diff
@@ -48,4 +48,13 @@ export default class AlleleCountColumnFormatter
         );
     }
+    
+    public static getReads(sample:Mutation, dataField:string): string
```
This isn't well-named, this function could really be summarized as `(sample ? sample[dataField] : '')`. It doesn't have anything to do with Reads

adamabeshouse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -25,7 +25,7 @@ export type Column<T> = {
     sortBy?:((data:T)=>(number|null)) | ((data:T)=>(string|null)) | ((data:T)=>(number|null)[]) | ((data:T)=>(string|null)[]);
     render:(data:T)=>JSX.Element;
-    download?:(data:T)=>string;
+    download?:(data:T, element?:{})=>string;
```
I don't see the need for this change

pieterlukasse commented:
Approval based on code (I did no manual tests).

adamabeshouse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -220,20 +249,27 @@ class LazyMobXTableStore<T> {
+            
+          //rowDownloadData is list of lists iterate over each element (transpose)
+            let processedRowsDownloadData = rowDownloadData[0].map(function(col:string, i:number) { 
+              return rowDownloadData.map(function(row) { 
```
row should be named col

adamabeshouse replied:
please rename this, its confusing

adamabeshouse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -220,20 +249,27 @@ class LazyMobXTableStore<T> {
-            tableDownloadData.push(rowDownloadData);
+            
+          //rowDownloadData is list of lists iterate over each element (transpose)
+            let processedRowsDownloadData = rowDownloadData[0].map(function(col:string, i:number) { 
```
col should be named row

adamabeshouse replied:
please do rename this, it's confusing

adamabeshouse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -220,20 +249,27 @@ class LazyMobXTableStore<T> {
         // add rows (including hidden columns)
         this.dataStore.sortedData.forEach((rowData:T) => {
-            const rowDownloadData:string[] = [];
+            const rowDownloadData:any[] = []; //string[]
```
rowDownloadData at index i is the output for column i

adamabeshouse replied:
the type is string[][]

adamabeshouse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -134,6 +134,35 @@ export function lazyMobXTableSort<T>(data:T[], metric:SortMetric<T>, ascending:b
+    let length = 1;
+    if (data instanceof Array) {
+        if (!(rowData instanceof Array) || rowData.length != data.length) {
+            throw new Error('Unexpected error!');
```
This is not good because it makes assumptions about the form of rowData. For example, in some applications we don't need the output of download to be the same length as the input rowData.

pieterlukasse replied:
@adamabeshouse "unique per row" indeed does not necessarily need to be the length of rowData. But the assumption of "same order" will remain. This means that in the example where output of download is length 2 while rowData is 3 we would map the first and second rows to the first and second output of download data while mapping the third row of rowData to some empty string. 

Although I agree this is more flexible, I tend to prefer a more defensive implementation for such parts. I.e. with such a flexible implementation, my question then would be: which part of the code is responsible for checking the integrity of the data? I mean, how do you avoid the scenario where the 2 items of the output of download should actually be bound to rows 2 and 3 instead of 1 and 2? An implementation that is too flexible will just silently map things incorrectly.

Maybe it would be a good idea if you could share your feedback in the form of `code` for this part? Perhaps an alternative PR that achieves the same so we can more easily compare?

adamabeshouse replied:
If the user needs it to go to 2 and 3 then they can pad the beginning with an empty string. As I see it the issue you're bringing up is that if it's used improperly there won't be an easy way to see that, which is valid. In that case, we can, as you say, require that column.download(rowData) is the same length for all columns, without peeking into rowData itself at all.

pieterlukasse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -134,6 +134,35 @@ export function lazyMobXTableSort<T>(data:T[], metric:SortMetric<T>, ascending:b
+    let length = 1;
+    if (data instanceof Array) {
+        if (!(rowData instanceof Array) || rowData.length != data.length) {
+            throw new Error('Unexpected error!');
```
@adamabeshouse we can make it a bit more flexible by allowing the output of download to be either 0, 1 or the same length as rowData. But having output of download be e.g. 2 while rowData is 3 would still be an error, right?

adamabeshouse replied:
I dont see why unique per row means it needs to be the length of rowData. You could end up outputting more than one row from a single rowData, or fewer rows than the length of rowData. That's the functional reason why I see this approach as wrong. From a code point of view, it violates one of the main rules of this library which is to be completely agnostic as to the form of rowData - once we start going down that road, it becomes less maintainable and general.

I don't even think the length needs to be the same across the row. You can just pad missing entries with the empty string, and build a number of rows equal to the max length column output.

alisman commented on src/pages/patientView/mutation/column/TumorColumnFormatter.tsx:
```diff
@@ -67,4 +66,17 @@ export default class TumorColumnFormatter {
+    public static getSample(data:Array<{sampleId:string}>): string|string[] {
+        let result =[];
+        if (data) {
+            for (let datum of data) {
```
i'm pretty sure for ... of is not supported in IE11.  there is a babel tranform for it but i don't think we have it installed and probably not worth it.  i would just use Array.forEach to be safe.  we definately already have a polyfill for that.

adamabeshouse commented on src/shared/components/lazyMobXTable/LazyMobXTable.tsx:
```diff
@@ -220,20 +252,28 @@ class LazyMobXTableStore<T> {
         this.dataStore.sortedData.forEach((rowData:T) => {
-            const rowDownloadData:string[] = [];
+            const rowDownloadData:any[] = []; //string[]
+            let refLength = getRefLength(this.columns, rowData);
```
okay it's almost there - the one remaining comment I have on this is the refLength approach is inefficient - you're generating the download data twice for each row. You could equivalently just use the length of the first column download - since we're not padding, we're requiring the same length, there's no need to find the max length to compare against.

adamabeshouse replied:
okay I just realized the other reason for finding the max - duplicating non-array download outputs. I think there should be a way to do this without generating all the download data twice.

oplantalech commented:
I've squashed all commits into one and I've rebased, but Travis is failing. Apparently some `*-docs.json` are out of sync. @adamabeshouse @alisman @onursumer @inodb do you know how to fix that?
